crypto/dilithium/p_dilithium3_asn1.c

```diff
@@ -0,0 +1,216 @@
+      !CBB_add_asn1(&algorithm, &oid, CBS_ASN1_OBJECT) ||
+      !CBB_add_bytes(&oid, dilithium3_asn1_meth.oid, dilithium3_asn1_meth.oid_len) ||
+      !CBB_add_asn1(&spki, &key_bitstring, CBS_ASN1_BITSTRING) ||
+      !CBB_add_u8(&key_bitstring, 0 /* padding */) ||
+      !CBB_add_bytes(&key_bitstring, key->pub, DILITHIUM3_PUBLIC_KEY_BYTES) ||
+      !CBB_flush(out)) {
+    OPENSSL_PUT_ERROR(EVP, EVP_R_ENCODE_ERROR);
+    return 0;
+  }
+
+  return 1;
+}
+
+static int dilithium3_pub_cmp(const EVP_PKEY *a, const EVP_PKEY *b) {
+  const DILITHIUM3_KEY *a_key = a->pkey.ptr;
+  const DILITHIUM3_KEY *b_key = b->pkey.ptr;
+  return OPENSSL_memcmp(a_key->pub, b_key->pub, DILITHIUM3_PUBLIC_KEY_BYTES) == 0;
+}
+
+static int dilithium3_priv_decode(EVP_PKEY *out, CBS *params, CBS *key, CBS *pubkey) {
+  // See https://datatracker.ietf.org/doc/draft-ietf-lamps-dilithium-certificates/ section 6.
+
+  // The parameters must be omitted.
+  if (CBS_len(params) != 0 ) {
+    OPENSSL_PUT_ERROR(EVP, EVP_R_DECODE_ERROR);
+    return 0;
+  }
+
+  return dilithium3_set_priv_raw(out, CBS_data(key), CBS_len(key), NULL, 0);
+}
+
+static int dilithium3_priv_encode(CBB *out, const EVP_PKEY *pkey) {
+  DILITHIUM3_KEY *key = pkey->pkey.ptr;
+  if (!key->has_private) {
+    OPENSSL_PUT_ERROR(EVP, EVP_R_NOT_A_PRIVATE_KEY);
+    return 0;
+  }
+  // See https://datatracker.ietf.org/doc/draft-ietf-lamps-dilithium-certificates/ section 6.
+  CBB pkcs8, algorithm, oid, private_key;
+  if (!CBB_add_asn1(out, &pkcs8, CBS_ASN1_SEQUENCE) ||
+      !CBB_add_asn1_uint64(&pkcs8, 0 /* version */) ||
+      !CBB_add_asn1(&pkcs8, &algorithm, CBS_ASN1_SEQUENCE) ||
+      !CBB_add_asn1(&algorithm, &oid, CBS_ASN1_OBJECT) ||
+      !CBB_add_bytes(&oid, dilithium3_asn1_meth.oid, dilithium3_asn1_meth.oid_len) ||
+      !CBB_add_asn1(&pkcs8, &private_key, CBS_ASN1_OCTETSTRING) ||
+      !CBB_add_bytes(&private_key, key->priv, DILITHIUM3_PRIVATE_KEY_BYTES) ||
+      !CBB_flush(out)) {
+    OPENSSL_PUT_ERROR(EVP, EVP_R_ENCODE_ERROR);
+    return 0;
+  }
+
+  return 1;
+}
+
+static int dilithium3_priv_encode_v2(CBB *out, const EVP_PKEY *pkey) {
+OPENSSL_PUT_ERROR(EVP, EVP_R_OPERATION_NOT_SUPPORTED_FOR_THIS_KEYTYPE);
+return 0;
+}
+
+static int dilithium3_size(const EVP_PKEY *pkey) {
+  return DILITHIUM3_SIGNATURE_BYTES;
+}
+
+static int dilithium3_bits(const EVP_PKEY *pkey) {
+  return 8 * (DILITHIUM3_PUBLIC_KEY_BYTES);
+}
+
+const EVP_PKEY_ASN1_METHOD dilithium3_asn1_meth = {
+    EVP_PKEY_DILITHIUM3,
+    // 1.3.6.1.4.1.2.267.7.6.5 = Dilithium SIG Round-3. These are temp values from
+    // https://github.com/IETF-Hackathon/pqc-certificates/blob/master/docs/oid_mapping.md
+    // as we await NIST to release OIDs.
+    {0x2B, 0x06, 0x01, 0x04, 0x01, 0x02, 0x82, 0x0B, 0x07, 0x06, 0x05},
+    11,
+    dilithium3_pub_decode,
+    dilithium3_pub_encode,
+    dilithium3_pub_cmp,
+    dilithium3_priv_decode,
+    dilithium3_priv_encode,
+    dilithium3_priv_encode_v2,
```

Comment:
It's caught in `EVP_marshal_private_key_v2()`. Leave it `NULL`.

Reply:
ah, nice. removed.